[BE] split seq_id to collective_seq_id and p2p_seq_id

c-p-i-o opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126733
* #126726
* #126581
* __->__ #125727

Summary:
Split out `seq_id` into `collective_seq_id` and `p2p_seq_id`. The main idea here is that collectives that go to all machines should have identical `collective_seq_id` and therefore it makes it easier to spot if one of machines isn't handling a collective operation.
Next, we can attempt to match up p2p operations to ensure that the sender(s)/receivers(s) are in sync.

Resolves issue: https://github.com/pytorch/pytorch/issues/125173

Test Plan:
Unit tests.

Reviewers:

Subscribers:

Tasks:

Tags:

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

c-p-i-o committed:
[WIP] rename seq_id to collective_seq_id

Summary:
Attempt to separate out collectives and P2P for debug purposes.
Pass isP2P to record.

Test Plan:
Unit tests.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

c-p-i-o committed:
Update on "[WIP] rename seq_id to collective_seq_id"


Summary:
Attempt to separate out collectives and P2P for debug purposes.
Pass isP2P to record.

Test Plan:
Unit tests.

Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

c-p-i-o committed:
Update on "[WIP] rename seq_id to collective_seq_id"


Summary:
Attempt to separate out collectives and P2P for debug purposes.
Pass isP2P to record.

Test Plan:
Unit tests.

Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wconstab commented:
this mostly lgtm, i didn't review carefully

one thought is whether we want to (correctly) bump version major, or, just keep old seq_id and add new ones and later deprecate it.

I think its worth checking if anyone's gonna be broken by this, but since not many users have developed scripts yet its probably fine to just bump the major and drop the old key so we have less baggage.  anyone else have a thought on that?

c-p-i-o committed:
Update on "[WIP] rename seq_id to collective_seq_id"


Summary:
Attempt to separate out collectives and P2P for debug purposes.
Pass isP2P to record.

Test Plan:
Unit tests.

Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -479,13 +490,15 @@ struct NCCLTraceBuffer {
       Event* start,
-      Event* end) {
+      Event* end,
+      bool isP2P = "false") {
```
`bool isP2P = false`, otherwise by default this is setting it to true. Probably better not to use a default argument here and just update the call sites.

c-p-i-o replied:
Good suggestion. Fixed!

zdevito reviewed (approved):
Looks good, thanks!

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o commented:
> one thought is whether we want to (correctly) bump version major, or, just keep old seq_id and add new ones and later deprecate it.

@wconstab - do you have a preference? Shall we keep the old `seq_id`? How do I find out I won't break anyone?

zdevito commented:
I'd just bump the major version number. We are still iterating on what needs to go into this and the fix for any manual scripts is easy.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2431,7 +2440,11 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
   // Bump collective counter
-  seq_++;
+  if (isP2POp(opType)) {
+    seqP2P_++;
```
is this path actually possible?  (to have a p2p op routed into the `ProcessGroupNCCL::collective` function)

c-p-i-o replied:
Not possible! Good catch. Fixed.

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -428,11 +435,12 @@ struct NCCLTraceBuffer {
-    // group)
-    size_t seq_id_;
+    // Both collective_seq_id_ and op_id_ are per_pg incrementing counters
+    // collective_seq_id refers to actual kernel launches (e.g. 1 per coalesced
```
collective_seq_id and p2p_seq_id refer to actual kernel launches (e.g. 1 per coalesced..), however collective_seq_id only increments for true collective operations (over all ranks in the group) while p2p_seq_id increments only over non-collective operations in the group.

c-p-i-o replied:
Fixed the comment. Thanks.

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2239,14 +2240,15 @@ c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
       rank,
       opType,
-      seq_,
+      seqCollective_,
```
Will WorkNCCL only be collective work, not P2P work?

c-p-i-o replied:
> does it make sense to pass seqCollective_ to seqP2P_ NCCL work?

That's a good point @shuqiangzhang!

@wconstab / @zdevito - do we need to differentiate p2p v/s collectives at the NCCL layer? I think the answer is yes - but wanted your thoughts.

I'll address this in a follow on pull request

wconstab replied:
WorkNCCL is not part of the 'NCCL layer' (as in, inside nccl library), it's part of pytorch pg-nccl layer still.  It's just a container we use to represent that there is a work object that's being tracked by our watchdog.  It should match up with our flight recorder.  When the watchdog reports a timeout of some operation at seq_id blah, that should correspond with FR dumps.

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1,4 +1,5 @@
 
+#include <torch/csrc/distributed/c10d/Work.hpp>
```
nit: is this needed? as ProcessGroupNCCL.h somehow already has access to Work?

c-p-i-o replied:
This was suggested by `visual studio`.  I removed it!

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o commented:
@pytorchbot merge

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o commented:
@pytorchbot merge
